flyio/caveats.md

```diff
@@ -0,0 +1,405 @@
+
+Macaroons are tokens. They are define in https://github.com/superfly/macaroon.
+At fly we pass them to services in `Authorization: Bearer <token>` HTTP headers.
+Several tokens can be passed in a single header, separated by commas. Access
+control is allowed if any one of the bearer tokens allows access.
+
+A Macaroon is extensible. Users that have a token can construct new tokens
+from them by adding Caveats. A Caveat is a restriction on the token's use.
+A token with no Caveats is all powerful, and tokens with Caveats wittle away
+at that power. Fly always issues tokens with an Org caveat which limits the
+token to operate on a single organization. A user in many organizations may
+pass in many tokens to operate across many organizations.
+
+Macaroons support "third party" caveats. These are special in a few ways.
+Third party caveats are "discharged" by an accompanying third party discharge
+token. A third party caveat in a token fails if there is no accompanying third party
+discharge passed along with the token. Fly only honors third party discharge tokens
+issued by certain trusted third party issuers. These are:
+- https://api.fly.io/aaa/v1
+- https://auth.fly.io
+
+Third party discharge tokens are also special in that we allow them to carry
+attestations, and that they must be "finalized".  These attestations are used
+to state a user's identity, such as the Fly user ID, or an SSO identity.
+A finalized token cannot be extended.
+
+## Access Control
+
+Tokens grant access and are logically a predicate on access requests.
+When evaluating if an operation can be performed, Fly services build up
+an access request and evaluate it against the bearer tokens. The access
+control predicate returns an error explaining an access control failure if
+access is not allowed. It returns no error if access is allowed.
+
+Fly uses the `flyio.Access` structure to represent access requests:
+
+```
+type Access struct {
+        Action         resset.Action `json:"action,omitempty"`
+        OrgID          *uint64       `json:"orgid,omitempty"`
+        AppID          *uint64       `json:"appid,omitempty"`
+        Feature        *string       `json:"feature,omitempty"`
+        Volume         *string       `json:"volume,omitempty"`
+        Machine        *string       `json:"machine,omitempty"`
+        MachineFeature *string       `json:"machine_feature,omitempty"`
+        Mutation       *string       `json:"mutation,omitempty"`
+        SourceMachine  *string       `json:"sourceMachine,omitempty"`
+        Cluster        *string       `json:"cluster,omitempty"`
+        Command        []string      `json:"command,omitempty"`
+}
+```
+
+The `Action` field encodes a set of actions requested: Read, Write, Create, Delete, or Control, encoded
+as a string "rwcdC". The special encoding "*" denotes the set of all possible actions.
+
+The other fields are optional fields, and denote what is being operated on.
```

Comment:
Worth noting that all parent resources must be specified. E.g. if a machine is specified, you must also specify an app and an org.

Reply:
added some docs around this.